src/plugins/WebOfTrust/util/TickerDelayedBackgroundJob.java

```diff
@@ -0,0 +1,363 @@
+        if (newExecutionTime < nextExecutionTime) {
+            nextExecutionTime = newExecutionTime;
+            if (state == JobState.RUNNING) {
+                // Will automatically schedule this run when the running job finishes.
+                return;
+            }
+            if (state == JobState.WAITING) {
+                // Best-effort attempt at removing the stale job to be replaced.
+                ticker.removeQueuedJob(waitingTickerJob);
+                // Replace the ticker job in case the above fails, so the stale job will not run.
+                waitingTickerJob = createTickerJob();
+            }
+            enqueueWaitingTickerJob(delay);
+        }
+    }
+
+    /**
+     * Schedules the ticker job (which will submit the background job to the executor) at the
+     * ticker with the given delay, or executes the ticker job immediately if the delay is zero.
+     * If the current state is {@code IDLE}, we the state is changed to {@code WAITING} and a new
+     * ticker job is created.
+     * Caller must ensure synchronization on {@code this}.
+     * @param delay the delay in ms
+     */
+    private void enqueueWaitingTickerJob(long delay) {
+        assert(state == JobState.IDLE || state == JobState.WAITING) :
+                "enqueueing ticker job in non-IDLE and non-WAITING state";
+        // Use a unique job for each (re)scheduling to avoid running twice.
+        if (state == JobState.IDLE) {
+            toWAITING();
+        }
+        if (delay > 0) {
+            ticker.queueTimedJob(waitingTickerJob, name + " (waiting)", delay, true, false);
+        } else {
+            waitingTickerJob.run();
+        }
+    }
+
+    /**
+     * Creates a tiny job that instructs the executor to run the background job immediately, to
+     * be executed by the ticker.
+     */
+    private Runnable createTickerJob() {
+        return new FastRunnable() {
+            @Override
+            public void run() {
+                synchronized(TickerDelayedBackgroundJob.this) {
+                    // If this runnable is not the waitingTickerJob, it has been rescheduled. Only
+                    // run if this is the expected ticker job, otherwise the expected job will run
+                    // real soon.
+                    if (this == waitingTickerJob) {
+                        executor.execute(realJob, name + " (running)");
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * {RUNNING} -> IDLE
+     */
+    private void toIDLE() {
+        assert(state == JobState.RUNNING) : "going to IDLE from non-RUNNING state";
+        assert(thread == Thread.currentThread()) : "going to IDLE from non-job thread";
+        thread = null;
+        state = JobState.IDLE;
+    }
+
+    /**
+     * {IDLE} -> WAITING
+     */
+    private void toWAITING() {
+        assert(state == JobState.IDLE) : "going to WAITING from non-IDLE state";
```

Comment:
You could add an assert(nextExecutionTime >= System.currentTimeMilis() && nextExecutionTime != NO_EXECUTION)